lib/Basic/Subprocess.cpp

```diff
@@ -436,17 +452,47 @@ void llbuild::basic::spawnProcess(
 
     // If we have been cancelled since we started, do nothing.
     if (!wasCancelled) {
-      int result = posix_spawn(&pid, args[0], /*file_actions=*/&fileActions,
-                               /*attrp=*/&attributes, const_cast<char**>(args.data()),
-                               const_cast<char* const*>(environment.getEnvp()));
+      int result = 0;
+
+#ifdef __APPLE__
+      int oldCWD = -1;
+      int chdirErr = 0;
+      if (attr.workingDir.size()) {
+        oldCWD = open(".", O_RDONLY);
+        if (pthread_fchdir_np(-1) == -1 && oldCWD != -1) {
+          // no existing per-thread cwd
+          close(oldCWD); oldCWD = -1;
+        }
+        if (-1 == pthread_chdir_np(attr.workingDir.data())) {
+          chdirErr = errno;
+        }
+      }
+      if (chdirErr != 0) {
+        result = -1;
+      }
+#endif
+
+      if (result == 0) {
+        result = posix_spawn(&pid, args[0], /*file_actions=*/&fileActions,
+                             /*attrp=*/&attributes, const_cast<char**>(args.data()),
+                             const_cast<char* const*>(environment.getEnvp()));
+      }
+
+#ifdef __APPLE__
+      if (attr.workingDir.size()) {
+        pthread_fchdir_np(oldCWD);
+        if (oldCWD != -1) close(oldCWD);
```

Comment:
presumably doesn't make sense to try to chdir to -1.

Reply:
Per the “docs” setting -1 clears the per thread working dir, restoring it to the process level.